src/coreclr/jit/importercalls.cpp

```diff
@@ -7558,15 +7558,10 @@ CORINFO_CLASS_HANDLE Compiler::impGetSpecialIntrinsicExactReturnType(GenTreeCall
 
             // Lookup can incorrect when we have __Canon as it won't appear
             // to implement any interface types.
-            //
-            // And if we do not have a final type, devirt & inlining is
-            // unlikely to result in much simplification.
-            //
-            // We can use CORINFO_FLG_FINAL to screen out both of these cases.
             const DWORD typeAttribs = info.compCompHnd->getClassAttribs(typeHnd);
-            bool        isFinalType = ((typeAttribs & CORINFO_FLG_FINAL) != 0);
+            bool        isShared    = ((typeAttribs & CORINFO_FLG_SHAREDINST) != 0);
```

Comment:
Are there any cases that get regressed by this change?

In theory, devirtualization can work for some cases of shared types (like `EqualityComparer<MyStruct<string>>`). I assume that we are not able to devirtualization like this today and it is what makes this change acceptable?

Reply:
The `CORINFO_FLG_FINAL` check already blocked shared generics, I just unblocked it for non shared, non final ones. AFAIR the reason for blocking it here is cause the VM would return the wrong comparer type in some cases.

Reply:
If you had a type like this:
```
struct MyStruct<T>
{
}
```
Both `CORINFO_FLG_FINAL` and `CORINFO_FLG_SHAREDINST` should be set for `MyStruct<string>`. So the original check did not block this type, but the new check does block it.

Reply:
> So the original check did not block this type

The comment suggested otherwise:
```C#
            // Lookup can incorrect when we have __Canon as it won't appear
            // to implement any interface types.
            //
            // And if we do not have a final type, devirt & inlining is
            // unlikely to result in much simplification.
            //
            // We can use CORINFO_FLG_FINAL to screen out both of these cases.
```
But even if so, I don't thing the VM can accurately check for things like `IEnumerable<T>` on shared generics.

Reply:
> Are there any cases that get regressed by this change?

And I'd need to check jit-diff to be sure since SPMI shows none, probably due to missed contexts.

Reply:
> The comment suggested otherwise:

The comment talks about `__Canon`. I am talking about `MyStruct<__Canon>` that the comment does not apply to - `CORINFO_FLG_FINAL` should be true for  `MyStruct<__Canon>`.

Reply:
The mono changes look ok to me.

Reply:
> Are there any cases that get regressed by this change? 

@MichalPetryka I also raised a concern about it in Discord - probably makes sense to run jit-diff and check regressions if any?

Reply:
> Are there any cases that get regressed by this change?

As far as I can see with Preview 5, the `Equals` call seems to not be inlined (sometimes even with simple cases like `int`) so the diffs wouldn't be big, nevertheless I've changes the code to not block shared final types.